<p align = "center" draggable=”false” ><img src="https://user-images.githubusercontent.com/37101144/161836199-fdb0219d-0361-4988-bf26-48b0fad160a3.png" 
     width="200px"
     height="auto"/>
</p>

# <h1 align="center" id="heading">Sentiment Analysis of Reddit Data using Reddit API</h1>

In this live coding session, we leverage the Python Reddit API Wrapper (`PRAW`) to retrieve data from subreddits on [Reddit](https://www.reddit.com), and perform sentiment analysis using [`pipelines`](https://huggingface.co/docs/transformers/main_classes/pipelines) from [HuggingFace ( 🤗 the GitHub of Machine Learning )](https://techcrunch.com/2022/05/09/hugging-face-reaches-2-billion-valuation-to-build-the-github-of-machine-learning/), powered by [transformer](https://arxiv.org/pdf/1706.03762.pdf).

## Objectives

At the end of the session, you will 

- know how to work with APIs
- feel more comfortable navigating thru documentation, even inspecting the source code
- understand what a `pipeline` object is in HuggingFace
- perform sentiment analysis using `pipeline`
- run a python script in command line and get the results

## How to Submit

- At the end of each task, commit* the work into the repository you created before the assignment
- After completing all three tasks, make sure to push the notebook containing all code blocks and output cells to your repository you created before the assignment
- Submit the link to the notebook in Canvas

\***NEVER** commit a notebook displaying errors unless it is instructed otherwise. However, commit often; recall git ABC = **A**lways **B**e **C**ommitting.

## Tasks

### Task I: Instantiate a Reddit API Object

The first task is to instantiate a Reddit API object using [PRAW](https://praw.readthedocs.io/en/stable/), through which you will retrieve data. PRAW is a wrapper for [Reddit API](https://www.reddit.com/dev/api) that makes interacting with the Reddit API easier unless you are already an expert of [`requests`](https://docs.python-requests.org/en/latest/).

#### 1. Install packages

Please ensure you've ran all the cells in the `imports.ipynb`, located [here](https://github.com/FourthBrain/MLE-8/blob/main/assignments/week-3-analyze-sentiment-subreddit/imports.ipynb), to make sure you have all the required packages for today's assignment.

####  2. Create a new app on Reddit 

Create a new app on Reddit and save secret tokens; refer to [post in medium](https://towardsdatascience.com/how-to-use-the-reddit-api-in-python-5e05ddfd1e5c) for more details.

- Create a Reddit account if you don't have one, log into your account.
- To access the API, we need create an app. Slight updates, on the website, you need to navigate to `preference` > `app`, or click [this link](https://www.reddit.com/prefs/apps) and scroll all the way down. 
- Click to create a new app, fill in the **name**, choose `script`, fill in  **description** and **redirect uri** ( The redirect URI is where the user is sent after they've granted OAuth access to your application (more info [here](https://github.com/reddit-archive/reddit/wiki/OAuth2)) For our purpose, you can enter some random url, e.g., www.google.com; as shown below.


    <img src="https://miro.medium.com/max/700/1*lRBvxpIe8J2nZYJ6ucMgHA.png" width="500"/>
- Jot down `client_id` (left upper corner) and `client_secret` 

    NOTE: CLIENT_ID refers to 'personal use script" and CLIENT_SECRET to secret.
    
    <div>
    <img src="https://miro.medium.com/max/700/1*7cGAKth1PMrEf2sHcQWPoA.png" width="300"/>
    </div>

- Create `secrets_reddit.py` in the same directory with this notebook, fill in `client_id` and `secret_id` obtained from the last step. We will need to import those constants in the next step.
    ```
    REDDIT_API_CLIENT_ID = "client_id"
    REDDIT_API_CLIENT_SECRET = "secret_id"
    REDDIT_API_USER_AGENT = "any string except bot; ex. My User Agent"
    ```
- Add `secrets_reddit.py` to your `.gitignore` file if not already done. NEVER push credentials to a repo, private or public. 

#### 3. Instantiate a `Reddit` object

Now you are ready to create a read-only `Reddit` instance. Refer to [documentation](https://praw.readthedocs.io/en/stable/code_overview/reddit_instance.html) when necessary.

In [1]:
import praw
import secrets_reddit

# Create a Reddit object which allows us to interact with the Reddit API
reddit = praw.Reddit(
    # YOUR CODE HERE
    client_id = secrets_reddit.REDDIT_API_CLIENT_ID,
    client_secret = secrets_reddit.REDDIT_API_CLIENT_SECRET,
    user_agent = secrets_reddit.REDDIT_API_USER_AGENT
)

In [2]:
print(reddit) 

<details>
<summary>Expected output:</summary>   

```<praw.reddit.Reddit object at 0x10f8a0ac0>```
</details>

#### 4. Instantiate a `subreddit` object

Lastly, create a `subreddit` object for your favorite subreddit and inspect the object. The expected outputs you will see are from `r/machinelearning` unless otherwise specified.

In [3]:
# YOUR CODE HERE
subreddit = reddit.subreddit("machinelearning")

What is the display name of the subreddit?

In [4]:
# YOUR CODE HERE
print(subreddit.display_name)

machinelearning


<details>
<summary>Expected output:</summary>   

    machinelearning
</details>

How about its title, is it different from the display name?

In [5]:
# YOUR CODE HERE
print(subreddit.title)

Machine Learning


<details>
<summary>Expected output:</summary>   

    Machine Learning
</details>

Print out the description of the subreddit:

In [6]:
# YOUR CODE HERE
print(subreddit.description[:500])

**[Rules For Posts](https://www.reddit.com/r/MachineLearning/about/rules/)**
--------
+[Research](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AResearch)
--------
+[Discussion](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3ADiscussion)
--------
+[Project](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AProject)
--------
+[News](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict


<details>
<summary>Expected output:</summary>

    **[Rules For Posts](https://www.reddit.com/r/MachineLearning/about/rules/)**
    --------
    +[Research](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AResearch)
    --------
    +[Discussion](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3ADiscussion)
    --------
    +[Project](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict_sr=on&q=flair%3AProject)
    --------
    +[News](https://www.reddit.com/r/MachineLearning/search?sort=new&restrict
</details>

### Task II: Parse comments

#### 1. Top Posts of All Time

Find titles of top 10 posts of **all time** from your favorite subreddit. Refer to [Obtain Submission Instances from a Subreddit Section](https://praw.readthedocs.io/en/stable/getting_started/quick_start.html)) if necessary. Verify if the titles match what you read on Reddit.

In [7]:
# try run this line, what do you see? press q once you are done
?subreddit.top 

Signature:
subreddit.top(
    *,
    time_filter: str = 'all',
    **generator_kwargs: Union[str, int, Dict[str, str]],
) -> Iterator[Any]
Docstring:
Return a :class:`.ListingGenerator` for top items.

:param time_filter: Can be one of: ``"all"``, ``"day"``, ``"hour"``,
    ``"month"``, ``"week"``, or ``"year"`` (default: ``"all"``).

:raises: :py:class:`ValueError` if ``time_filter`` is invalid.

Additional keyword arguments are passed in the initialization of
:class:`.ListingGenerator`.

This method can be used like:

.. code-block:: python

    reddit.domain("imgur.com").top(time_filter="week")
    reddit.multireddit(redditor="samuraisam", name="programming").top(time_filter="day")
    reddit.redditor("spez").top(time_filter="month")
    reddit.redditor("spez").comments.top(time_filter="year")
    reddit.redditor("spez").submissions.top(time_filter="all")
    reddit.subreddit("all").top(time_filter="hour")
File:      ~/miniconda3/envs/sa/lib/python3.8/site-packages/praw/models/listi

In [8]:
# YOUR CODE HERE
top10 = subreddit.top(limit=10)
for submission in top10:
    print(submission.title)

[Project] From books to presentations in 10s with AR + ML
[D] A Demo from 1993 of 32-year-old Yann LeCun showing off the World's first Convolutional Network for Text Recognition
[R] First Order Motion Model applied to animate paintings
[N] AI can turn old photos into moving Images / Link is given in the comments - You can also turn your old photo like this
[D] This AI reveals how much time politicians stare at their phone at work
[D] Types of Machine Learning Papers
[D] The machine learning community has a toxicity problem
[Project] NEW PYTHON PACKAGE: Sync GAN Art to Music with "Lucid Sonic Dreams"! (Link in Comments)
I made a robot that punishes me if it detects that if I am procrastinating on my assignments [P]
[P] Using oil portraits and First Order Model to bring the paintings back to life


<details> <summary>Expected output:</summary>

    [Project] From books to presentations in 10s with AR + ML
    [D] A Demo from 1993 of 32-year-old Yann LeCun showing off the World's first Convolutional Network for Text Recognition
    [R] First Order Motion Model applied to animate paintings
    [N] AI can turn old photos into moving Images / Link is given in the comments - You can also turn your old photo like this
    [D] This AI reveals how much time politicians stare at their phone at work
    [D] Types of Machine Learning Papers
    [D] The machine learning community has a toxicity problem
    [Project] NEW PYTHON PACKAGE: Sync GAN Art to Music with "Lucid Sonic Dreams"! (Link in Comments)
    [P] Using oil portraits and First Order Model to bring the paintings back to life
    [D] Convolution Neural Network Visualization - Made with Unity 3D and lots of Code / source - stefsietz (IG)    
</details>

#### 2. Top 10 Posts of This Week

What are the titles of the top 10 posts of **this week** from your favorite subreddit?

In [9]:
# YOUR CODE HERE
top10 = subreddit.top(time_filter='week', limit=10)
for submission in top10:
    print(submission.title)

30% of Google's Reddit Emotions Dataset is Mislabeled [D]
[D] An accusation of academic misconduct by Prof. Yisen Wang (Peking University) in ICML2021 and NeurIPS2021
[R] mixed reality future — see the world through artistic lenses — made with NeRF
[N] First-Ever Course on Transformers: NOW PUBLIC
[D] Why are Corgi dogs so popular in machine learning (especially in the image generation community)?
[D] Noam Chomsky on LLMs and discussion of LeCun paper (MLST)
[P] Sioyek 1.4 | Academic PDF Viewer
[R] So someone actually peer-reviewed this and thought "yeah, looks good"?
[N] Andrej Karpathy is leaving Tesla
[D] How do you verify the novelty of your research?


<details><summary>Expected output:</summary>

    [N] Ian Goodfellow, Apple’s director of machine learning, is leaving the company due to its return to work policy. In a note to staff, he said “I believe strongly that more flexibility would have been the best policy for my team.” He was likely the company’s most cited ML expert.
    [R][P] Thin-Plate Spline Motion Model for Image Animation + Gradio Web Demo
    [P] I’ve been trying to understand the limits of some of the available machine learning models out there. Built an app that lets you try a mix of CLIP from Open AI + Apple’s version of MobileNet, and more directly on your phone's camera roll.
    [R] Meta is releasing a 175B parameter language model
    [N] Hugging Face raised $100M at $2B to double down on community, open-source & ethics
    [P] T-SNE to view and order your Spotify tracks
    [D] : HELP Finding a Book - A book written for Google Engineers about foundational Math to support ML
    [R] Scaled up CLIP-like model (~2B) shows 86% Zero-shot on Imagenet
    [D] Do you use NLTK or Spacy for text preprocessing?
    [D] Democratizing Diffusion Models - LDMs: High-Resolution Image Synthesis with Latent Diffusion Models, a 5-minute paper summary by Casual GAN Papers
</details>

💽❓ Data Question:

Check out what other attributes the `praw.models.Submission` class has in the [docs](https://praw.readthedocs.io/en/stable/code_overview/models/submission.html). 

1. After having a chance to look through the docs, is there any other information that you might want to extract? How might this additional data help you?

   * I might want to extract *author*, *name*, *comments* and *num_comments*, *score* and *upvote_ratio*.
   * Last information could tell us more about popularity of the post.

Write a sample piece of code below extracting three additional pieces of information from the submission below.

In [10]:
# YOUR CODE HERE
top10 = subreddit.top(time_filter='week', limit=10)
for submission in top10:
    title = submission.title
    comments = submission.num_comments
    score = submission.score
    upvote_ratio = submission.upvote_ratio
    print(f"{title} \n\t Comments: {comments}  Score: {score}  Upvote Ratio: {upvote_ratio}")

30% of Google's Reddit Emotions Dataset is Mislabeled [D] 
	 Comments: 111  Score: 668  Upvote Ratio: 0.98
[D] An accusation of academic misconduct by Prof. Yisen Wang (Peking University) in ICML2021 and NeurIPS2021 
	 Comments: 115  Score: 648  Upvote Ratio: 0.97
[R] mixed reality future — see the world through artistic lenses — made with NeRF 
	 Comments: 15  Score: 356  Upvote Ratio: 0.96
[N] First-Ever Course on Transformers: NOW PUBLIC 
	 Comments: 37  Score: 344  Upvote Ratio: 0.92
[D] Why are Corgi dogs so popular in machine learning (especially in the image generation community)? 
	 Comments: 68  Score: 320  Upvote Ratio: 0.92
[D] Noam Chomsky on LLMs and discussion of LeCun paper (MLST) 
	 Comments: 240  Score: 273  Upvote Ratio: 0.88
[P] Sioyek 1.4 | Academic PDF Viewer 
	 Comments: 16  Score: 248  Upvote Ratio: 1.0
[R] So someone actually peer-reviewed this and thought "yeah, looks good"? 
	 Comments: 77  Score: 227  Upvote Ratio: 0.96
[N] Andrej Karpathy is leaving Tesla 
	

💽❓ Data Question:

2. Is there any information available that might be a concern when it comes to Ethical Data?
  
   * I think that *over_18* might be a concern.

#### 3. Comment Code

Add comments to the code block below to describe what each line of the code does (Refer to [Obtain Comment Instances Section](https://praw.readthedocs.io/en/stable/getting_started/quick_start.html) when necessary). The code is adapted from [this tutorial](https://praw.readthedocs.io/en/stable/tutorials/comments.html)

The purpose is 
1. to understand what the code is doing 
2. start to comment your code whenever it is not self-explantory if you have not (others will thank you, YOU will thank you later 😊) 

In [11]:
%%time
from praw.models import MoreComments

# YOUR COMMENT HERE
# Initialize an empty list
top_comments = []

# YOUR COMMENT HERE
# Iterate through top 10 posts
for submission in subreddit.top(limit=10):
    # YOUR COMMENT HERE
    # Iterate through top-level comments in the thread
    for top_level_comment in submission.comments:
        # YOUR COMMENT HERE
        # If a comment  is an instance of MoreComments, ignore it
        if isinstance(top_level_comment, MoreComments):
            continue
        # YOUR COMMENT HERE
        # Otherwise, append the list with comment's body
        top_comments.append(top_level_comment.body)

CPU times: user 468 ms, sys: 36.4 ms, total: 505 ms
Wall time: 1min 52s


#### 4. Inspect Comments

How many comments did you extract from the last step? Examine a few comments. 

In [12]:
#YOUR CODE HERE  # the answer may vary 693 for r/machinelearning
len(top_comments)

737

In [17]:
import random

[random.choice(top_comments) for i in range(3)]

['Take my money', 'Need to know the pr0n ratio', 'Wow.']

<details> <summary>Some of the comments from `r/machinelearning` subreddit are:</summary>

    ['Awesome visualisation',
    'Similar to a stack or connected neurons.',
    'Will this Turing pass the Turing Test?']
</details>

💽❓ Data Question:

3. After having a chance to review a few samples of 5 comments from the subreddit, what can you say about the data? 

   * The data is not very clean. There are many acronyms, abbreviations or chat slang words (lol, plz, ccp, ...). Not sure how sentiment analysis models are good with these kind of words.
   * I also noticed some incorrect word contractions like dont and cant.
   * Our task is to do sentiment analysis of Tesla stocks, but in the data we can not see anything about Tesla or any other stocks.


HINT: Think about the "cleanliness" of the data, the content of the data, think about what you're trying to do - how does this data line up with your goal?

#### 5. Extract Top Level Comment from Subreddit `TSLA`.

Write your code to extract top level comments from the top 10 topics of a time period, e.g., year, from subreddit `TSLA` and store them in a list `top_comments_tsla`.  

In [19]:
%%time
# YOUR CODE HERE
subreddit = reddit.subreddit("TSLA")

top10 = subreddit.top(time_filter='year', limit=10)
for submission in top10:
    print(submission.title)

from praw.models import MoreComments

# Initialize an empty list
top_comments_tsla = []

# Iterate through top 10 posts
for submission in subreddit.top(limit=10):
    # Iterate through top-level comments in the thread
    for top_level_comment in submission.comments:
        # If a comment  is an instance of MoreComments, ignore it
        if isinstance(top_level_comment, MoreComments):
            continue
        # Otherwise, append the list with comment's body
        top_comments_tsla.append(top_level_comment.body)
    

Up Vote if you would like to see a 1for 5 stock split?
Raise your hands if you stayed long and didn’t de-risk on Tesla because of your research and faith in the actual company.
Vote up for a 10 for 1 stock split (at 1500/share before split)!
I’m a small fry, but happy to say I’ve just passed Half a share of TSLA! Hoping to get to a full share by year end.
Record Deliveries in Q4 2021
All in on TSLA
Pepsi buying Tesla trucks
Tesla To 1.2k and Beyond
Bought on the dip today. Who’s with me?
Congrats my fellow longs
CPU times: user 338 ms, sys: 0 ns, total: 338 ms
Wall time: 53.3 s


In [20]:
len(top_comments_tsla) # Expected: 174 for r/machinelearning

170

In [21]:
[random.choice(top_comments_tsla) for i in range(3)]

['Base in uk, who knows best apps to buy and sell or FOR investors for TSLA.....Thank you guys!',
 'What does a split mean? Is it bad for people that have shares?',
 'Stacked up on Tesla and big oil 2 years ago. Can’t be happier.']

<details>
<summary>Some of the comments from `r/TSLA` subreddit:</summary>

    ['I bought puts',
    '100%',
    'Yes. And I’m bag holding 1200 calls for Friday and am close to throwing myself out the window']
</details>

💽❓ Data Question:

4. Now that you've had a chance to review another subreddits comments, do you see any differences in the kinds of comments either subreddit has - and how might this relate to bias?

   * Comments from TSLA subreddit fit much better our task than comments from the machinelearning subreddit. 
   * Our population should consist from people who discuss and display their feelings about Tesla. So, we want to extract data from such communities. 
   * Maybe we are in a danger to fall in a selection bias here. But what a result we could get from communities that are not talking about Tesla? 
   * To avoid selection bias and availability bias too, we should include more communities which users are displaying sentiments toward Tesla.
   * By choosing additional communities we shouild try to avoid other types of bias as well..



### Task III: Sentiment Analysis

Let us analyze the sentiment of comments scraped from `r/TSLA` using a pre-trained HuggingFace model to make the inference. Take a [Quick tour](https://huggingface.co/docs/transformers/quicktour). 

#### 1. Import `pipeline`

In [22]:
from transformers import pipeline # YOUR CODE HERE

#### 2. Create a Pipeline to Perform Task "sentiment-analysis"

In [23]:
sentiment_model = pipeline("sentiment-analysis") # YOUR CODE HERE

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english)


#### 3. Get one comment from list `top_comments_tsla` from Task II - 5.

In [28]:
random.seed(11)
comment = random.choice(top_comments_tsla)

In [29]:
comment

"$580 and falling today. What's your price target?"

The example comment is: `'Bury Burry!!!!!'`. Print out what you get. For reproducibility, use the same comment in the next step; consider setting a seed.

#### 4. Make Inference!

In [37]:
sentiment = sentiment_model(comment)  # YOUR CODE HERE 
print(type(sentiment))
sentiment

<class 'list'>


[{'label': 'NEGATIVE', 'score': 0.9976972937583923}]

What is the type of the output `sentiment`?

```
YOUR ANSWER HERE

```
   * It is a list of dictionaries.


In [32]:
print(f'The comment: {comment}')
print(f'Predicted Label is {sentiment[0]["label"]} and the score is {sentiment[0]["score"]:.3f}')

The comment: $580 and falling today. What's your price target?
Predicted Label is NEGATIVE and the score is 0.998


For the example comment, the output is:

    The comment: Bury Burry!!!!!
    Predicted Label is NEGATIVE and the score is 0.989

🖥️❓ Model Question:

1. What does the score represent?
   * The score represents probability for the predicted label.

### Task IV: Put All Together

Let's pull all the piece together, create a simple script that does 

- get the subreddit
- get comments from the top posts for given subreddit
- run sentiment analysis 

#### Complete the Script

Once you complete the code, running the following block writes the code into a new Python script and saves it as `top_tlsa_comment_sentiment.py` under the same directory with the notebook. 

In [38]:
%%writefile top_tlsa_comment_sentiment.py

import secrets_reddit
import random

from typing import Dict, List

from praw import Reddit
from praw.models.reddit.subreddit import Subreddit
from praw.models import MoreComments

from transformers import pipeline


def get_subreddit(display_name:str) -> Subreddit:
    """Get subreddit object from display name

    Args:
        display_name (str): Name of subreddit (community)

    Returns:
        Subreddit: Subreddit object
    """
    reddit = Reddit(
        client_id=secrets_reddit.REDDIT_API_CLIENT_ID,        
        client_secret=secrets_reddit.REDDIT_API_CLIENT_SECRET,
        user_agent=secrets_reddit.REDDIT_API_USER_AGENT
        )
    
    subreddit = reddit.subreddit(display_name) # YOUR CODE HERE
    return subreddit

def get_comments(subreddit:Subreddit, limit:int=3) -> List[str]:
    """ Get comments from subreddit

    Args:
        subreddit (Subreddit): Subreddit object
        limit (int, optional): Number of posts (submissions) to extract. Defaults to 3.

    Returns:
        List[str]: List of comments
    """
    top_comments = []
    for submission in subreddit.top(limit=limit):
        for top_level_comment in submission.comments:
            if isinstance(top_level_comment, MoreComments):
                continue
            top_comments.append(top_level_comment.body)
    return top_comments

def run_sentiment_analysis(comment:str) -> Dict:
    """Run sentiment analysis on comment using default distilbert model
    
    Args:
        comment (str): Comment
        
    Returns:
        str: Sentiment analysis result
    """
    sentiment_model = pipeline("sentiment-analysis")  # YOUR CODE HERE
    sentiment = sentiment_model(comment)
    return sentiment[0]


if __name__ == '__main__':
    subreddit = get_subreddit("TSLA")  # YOUR CODE HERE
    comments = get_comments(subreddit)
    comment = random.choice(comments)  # YOUR CODE HERE
    sentiment = run_sentiment_analysis(comment)
    
    print(f'The comment: {comment}')
    print(f'Predicted Label is {sentiment["label"]} and the score is {sentiment["score"]:.3f}')

Writing top_tlsa_comment_sentiment.py


Run the following block to see the output.

In [41]:
# To disable tokenizers parallelism warning
import os
os.environ["TOKENIZERS_PARALLELISM"] = "false"

!python top_tlsa_comment_sentiment.py

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english)
The comment: You did - nicely done !!!
Predicted Label is POSITIVE and the score is 1.000


<details><summary> Expected output:</summary>

    No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english)
    The comment: When is DOGE flying
    Predicted Label is POSITIVE and the score is 0.689
</details>

💽❓ Data Question:

5. Is the subreddit active? About how many posts or threads per day? How could you find this information?

     
  * This subreddit is active. We can list creation dates of newest posts or comments to confirm that.
  * Distribution of post per day was analyzed below Data Question 6.
    * In almost 96% of days there were 1 - 6 posts per day
    * 1 post was in about 43% of days and 2 posts in 23%
   * To get creation dates for posts I used the Submission object with the argument *created*
   * Counter collection was used twice to get distribution of posts per days
     1. count number of posts per day
     1. count number of days with particular number of postings
   * Similar analysis could be done with Pandas DataFrame

In [225]:
# import datetime
from datetime import datetime

# List creation dates for 5 newest posts
posts = subreddit.new( limit=5)
for submission in posts:
    print(datetime.fromtimestamp(submission.created))

2022-07-15 00:05:44
2022-07-14 04:21:43
2022-07-13 12:42:15
2022-07-13 11:17:53
2022-07-13 11:13:47


In [178]:
# List creation dates for 5 newest comments
comments = subreddit.comments( limit=5)
for comment in comments:
    print(datetime.fromtimestamp(comment.created))

2022-07-15 00:07:17
2022-07-15 00:05:54
2022-07-14 19:09:05
2022-07-14 11:14:57
2022-07-14 11:12:23


💽❓ Data Question:

6. Does there seem to be a large distribution of posters or a smaller concentration of posters who are very active? What kind of impact might this have on the data?

  * There is a large distribution of posters
    * Percentage of posters with 1 or 2 posts is about 87%.
    * There are 3 posters (of 482) with more than 20 posts
   * This is good for data because it is not impacted by a small group of posters. That helps to avoid selection bias.

#### Collecting authors and creation dates

In [227]:
# Collect information about authors and creation dates for every post in subreddit
authors = []
date_created = []

posts = subreddit.new( limit=None)
# Iteratr through all posts
for submission in posts:
    # To avoid issues with authors eqal None
    if submission.author is not None:
        authors.append(submission.author.name)
        date_created.append(datetime.fromtimestamp(submission.created).strftime('%Y-%m-%d'))

In [229]:
print(len(authors))
print(authors[:4])
print(len(date_created))
print(date_created[:4])

976
['wewewawa', 'AdmirableActuator', 'StocksWithCamden', 'ju-ju2020']
976
['2022-07-15', '2022-07-14', '2022-07-13', '2022-07-13']


#### Submissions per days

In [252]:
# Import Counter
from collections import Counter
# Count number of posts per dates
n_posts = Counter(date_created)
print(list(n_posts.items())[:4])

# Count frequency of postings and sort 
freq = Counter(n_posts.values()).most_common()
print('\n', freq)

[('2022-07-15', 1), ('2022-07-14', 1), ('2022-07-13', 4), ('2022-07-11', 1)]

 [(1, 164), (2, 87), (3, 45), (4, 37), (5, 17), (6, 12), (11, 3), (7, 3), (8, 3), (13, 2), (17, 2), (12, 1), (33, 1), (15, 1)]


* 164 days we have only one posting, 87 days 2 postings and so on. Let's calculate percentage of days with 1 - 6 postings.

In [253]:
# Calculate number of days with 1 - 6 postings
n = 6  
s = 0
for i in range(n):
    s += freq[i][1]
print(f"Total number of days with 1 - 6 postings per day: {s}")

# Total number of days 
d = len(set(date_created))
print(f"Total number of days: {d}")
print(f"Percentage of days with 1 - 6 postings: {round(100 * (s / d), 2)}%")
print(f"Percentage of days with 1 posting: {round(100 * (freq[0][1]/ d), 2)}%")
print(f"Percentage of days with 2 postings: {round(100 * (freq[1][1] / d), 2)}%")

Total number of days with 1 - 6 postings per day: 362
Total number of days: 378
Percentage of days with 1 - 6 postings: 95.77%
Percentage of days with 1 posting: 43.39%
Percentage of days with 2 postings: 23.02%


#### Submissions per authors

In [254]:
# Number of posts per authors
n_authors = Counter(authors)
print(list(n_authors.items())[:4])

# Count frequency of postings and sort 
afreq = Counter(n_authors.values()).most_common()
print('\n', afreq)

[('wewewawa', 52), ('AdmirableActuator', 1), ('StocksWithCamden', 1), ('ju-ju2020', 2)]

 [(1, 348), (2, 71), (3, 22), (6, 10), (5, 6), (9, 5), (4, 5), (8, 2), (11, 2), (14, 2), (52, 1), (51, 1), (13, 1), (12, 1), (7, 1), (16, 1), (10, 1), (23, 1), (15, 1)]


In [265]:
# Calculate number of authors with 1 - 2 postings
n = 2 
nl = 20
s = 0
sl = 0
for i in range(len(afreq)):
    # Calculate number of authors with only few posts
    if afreq[i][0] <= 2:
        s += afreq[i][1]
    # Calculate number of authors with many posts       
    elif afreq[i][0] >= nl:
        sl += afreq[i][1]
print(f"Number of authors with 1 - 2 postings per day: {s}")
print(f"Number of authors with more than {nl} postings per day: {sl}")

# Total number of authors 
a = len(set(authors))
print(f"Total number of authors: {a}")
print(f"Percentage of authors with 1 - 2 postings: {round(100 * (s / a), 2)}%")


Number of authors with 1 - 2 postings per day: 419
Number of authors with more than 20 postings per day: 3
Total number of authors: 482
Percentage of authors with 1 - 2 postings: 86.93%
